Added Audio Input Support

bushrat011899 opened the pull request:
# Objective

- Fixes #9954

## Solution

- Created a new feature in `bevy_audio`, `input`, which listens to the default audio input device (if any) and publishes samples into the ECS using `Event`s.
- Added an example of a simple oscilloscope, which draws the waveform of every channel in the recorded audio input using `Gizmos`

---

## Changelog

- Added `input` feature to `bevy_audio`, which is activated by the `audio-input` feature in `bevy`/`bevy_internal`. This is an optional feature. With it disabled, everything added in this PR is removed, leaving no impact.
- Added a `NonSend` `Resource` `AudioInputStream`, which handles the `cpal` `Stream`. If this resource is dropped, then the `Stream` is also dropped with no memory leaked.
- Added an `Event` `AudioInputEvent`, which is dispatched whenever an audio input is recorded by `AudioInputStream`.
- Added a simple system `handle_input_stream` which marshals `AudioInputEvent` events out of `AudioInputStream` and into the ECS.
- Updated `AudioPlugin` to setup everything required to start recording from the default audio input device.
- Added an example `audio/audio_input`, which is a simple oscilloscope rendering recorded audio samples using `Gizmos`. See below animated GIF for a low-framerate representation of the example.
- Added some commands and control starting and stopping the audio input stream.
- Added `AudioInput` and `AudioInputOptions` to allow for the selection of a non-default audio input device and configuration (if desired). Alternatively, a `cpal` `Device` can be turned into a `AudioInput` which can then be used to create the `AudioInputStream`.

## Notes

With the `audio-input` feature enabled, all a user has to do to get access to the microphone is create a system which listens to `AudioInputEvent` events:

```rust
use bevy::prelude::*;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_systems(Update, (system, push_to_listen))
        .run();
}

/// Only record audio while the user has held the space-bar key
fn push_to_listen(mut commands: Commands, keyboard_input: Res<Input<KeyCode>>) {
    if keyboard_input.just_pressed(KeyCode::Space) {
        commands.start_recording_audio();
    } else if keyboard_input.just_released(KeyCode::Space) {
        commands.stop_recording_audio();
    }
}

fn system(mut inputs: EventReader<AudioInputEvent>) {
    for input in inputs.read() {
        // You now have access to the sampled input waveform and metadata
    }
}
```

`AudioInputEvent` has been designed to include all the contextual information a consumer might need to work with the recording (sample rate, channels, timing). This metadata does increase the size of this struct more than it might need to, as this information could live in a separate resource. However, I think this is more ergonomic, and considering the relative size of the audio sample itself (usually 900 `f32` values in a vector on my machine) compared to the metadata (less than 100 bytes), it's not substantially more wasteful.

I haven't tested latency or performance impact of this running alongside other Bevy modules. From my rough experimentation with the example created, it appears to be fairly low latency, but someone more qualified would be a better judge than me on that (I was a bad drummer). I have a [short slow-motion video](https://youtu.be/4J3-FRwfAe4) where I tap on my microphone (to create an impulse). Recording done on an Xperia 1 IV pointed at a Windows 10 PC. The video appears to show a latency less than 10 ms (the smallest measurement possible under this particular setup), which should be sufficient for communication applications (e.g., voice chat within a multiplayer game), but further testing would be required to assess more demanding applications (e.g., vocal effects for a live performance, etc.)

I have attempted to minimise/eliminate any allocation in the hot-path beyond Bevy's ECS operations using bounded MPSC channels and pre-allocated vectors, but I'm still new to allocation-free Rust programming, so there may be room for further optimisation going forward.

Finally, I'm adding this PR in a ready-to-go state, since it works and passes all my local CI, but I am open to suggestions on adding/removing parts, since this is a mostly brand new feature.

![audio_input Example](https://github.com/bevyengine/bevy/assets/2217286/8e394b0a-04da-4189-b52a-48d7d154c179)

mockersf commented:
> Another area of concern is this will start piping data from the audio input source into the ECS straight away, and continue to do so for the entire runtime of the application. This can be controlled by calling `pause`/`play` on the `Stream` contained within `AudioInputStream`, but it is worth noting here.

I think we shouldn't start recording at start, but only after it's asked for by the game

bushrat011899 commented:
> I think we shouldn't start recording at start, but only after it's asked for by the game

Totally fair. I've updated this PR to include some `Command`'s which control whether the stream is collecting input data. 

```rust
fn push_to_listen(mut commands: Commands, keyboard_input: Res<Input<KeyCode>>) {
    if keyboard_input.just_pressed(KeyCode::Space) {
        commands.start_recording_audio();
    } else if !keyboard_input.pressed(KeyCode::Space) {
        commands.stop_recording_audio();
    }
}
```

With this, I've removed the setup code I had before, so with the `audio-input` feature enabled, all that happens is the input device is found and a stream is setup, but it is no longer started.

jdm commented on examples/audio/audio_input.rs:
```diff
@@ -0,0 +1,72 @@
+fn push_to_listen(mut commands: Commands, keyboard_input: Res<Input<KeyCode>>) {
+    if keyboard_input.just_pressed(KeyCode::Space) {
+        commands.start_recording_audio();
+    } else if !keyboard_input.pressed(KeyCode::Space) {
```
```suggestion
    } else if keyboard_input.just_released(KeyCode::Space) {
```

jdm reviewed (approved):
I really like this! It's clean, has a simple interface, and the example shows that it fits in to bevy's model nicely.

SolarLiner commented on crates/bevy_audio/src/audio_input.rs:
```diff
@@ -0,0 +1,180 @@
+            .build_input_stream(
+                &config,
+                move |data: &[f32], info: &InputCallbackInfo| {
+                    let event = AudioInputEvent {
+                        samples: data.to_vec(),
+                        info: info.clone(),
+                        config: config_clone.clone(),
+                    };
+
+                    if let Err(error) = tx.send(event) {
+                        warn!("Audio Input Error: {}", error);
+                    };
```
Allocations in audio threads are discouraged. To learn more about why, I recommend this article: http://www.rossbencina.com/code/real-time-audio-programming-101-time-waits-for-nothing

In short, audio threads are supposed to never be paused while the data callback is executing. Because any system call (such as asking the OS for memory) can cause the OS to pause the thread, the callback might run for longer than the duration of an audio buffer, and the audio will start to glitch due to "buffer underrun/underflow", that is audio input data is being fed into the callback function faster than it can process it.

The callback here (potentially) asks for memory in two places: `data.to_vec()` and `tx.send(event)`, the former at each invocation of the callback, and the latter in cases where the channel is filling up faster than it is emptying.

In general, audio programming needs to be real-time safe, that is, wait-free (that is, progress is guaranteed) and lock-free or non-blocking (threads must not be paused). Only structures that have both these characteristics should be used here, and in this case, the solution is a circular buffer (eg. the `rtrb` crate), that is preallocated once and offers both writes and reads that are both wait-free and lock-free.

For this kind of application, the circular buffer can directly hold audio data, without the need for an extra `AudioInputEvent` struct, as the `info` and `config` are not going to change for the duration of the input stream, and the contiguous layouts will help with cache hit rates. For example with the `rtrb` crate, the producer and consumer can both write (resp. read) chunks of audio data directly into the buffer at once.

Using a circular buffer means that if it is full, the oldest data is dropped, which results in glitches as well, but the callback execution time is still bounded, and if the consumer of this audio data is running late, well that's nothing we can do about it and not our problem. 😆 

As for how big the circular buffer should be, a first pass could be set at a multiple of the input data size (something like 3 or 5 times the buffer, here we could define `let circ_buffer_size = 5 * config.buffer_size`), however for safety we could envison having a entire second of audio buffer available, so `let circ_buffer_size = config.sample_rate as usize`. Measurements on how much space is actually used could be carried out on different machines, and a 99.9% percentile could be used as the buffer size.

bushrat011899 replied:
Thanks for the detailed write-up on this! I really appreciate the amount of work this represents, and I'm deeply sorry I didn't respond to you sooner. I wanted to make sure I had a good chunk of free time to really take this all on-board, and I finally had some on the commute between some family functions this weekend.

What I've done to address this issue is two-fold:

1. Switched to using bounded `sync_channel`'s instead of unbounded `channel`'s.
2. Added a second channel for sending pre-allocated sample buffers between the audio input callback (the hot-path) and the rest of Bevy.

Since the `sync_channel` is bounded, it is backed by a fixed (and pre-allocated) array, so sending events will not block, and not allocate. To handle audio buffer vectors, I've created a second channel which `Bevy` can use to send pre-allocated buffers back to the audio system. This could be thought of as recycling the existing buffers, but to allow the event lifetime to be unrestricted, I'm just making new vectors within the event system 1:1 with sent events. To handle the case where my predicted buffer size is inadequate, the audio hot path _can_ allocate to extend the vector it is provided, and the new length will be used for further pre-allocated buffers. As long as the audio system sends buffers of constant size (which `cpal` is supposed to do), they _shouldn't_ allocate in the hot path.

While I agree a ring buffer would be the more direct solution to this problem, this channel based alternative has the same end result, with the added benefit of allowing re-allocation if it has to happen, and only relying on standard library primitives.

I personally haven't performance tested this implementation, but it does work for the provided example, and appears to be low enough latency for my intended use case (voice-chat).

SolarLiner commented on crates/bevy_audio/src/audio_input.rs:
```diff
@@ -0,0 +1,180 @@
+    }
+
+    /// Start recording audio from this input.
+    pub fn start(&self) -> Result<(), PlayStreamError> {
+        self.stream.play()
+    }
+
+    /// Stop recording audio from this input.
+    pub fn stop(&self) -> Result<(), PauseStreamError> {
+        self.stream.pause()
+    }
```
We should probably wrap the error type, to prevent "leaking the internals" and allow us to change the underlying implementation without introducing breaking changes

bushrat011899 replied:
Agreed!

SolarLiner reviewed (changes requested):
Good idea on the implementation, some considerations to be had for the actual audio processing and communication, as audio programming is a type of real-time programming, which means additional rules and restrictions on the "hot path".

bushrat011899 commented on crates/bevy_audio/src/audio_input.rs:
```diff
@@ -0,0 +1,180 @@
+    pub fn try_default() -> Option<Self> {
+        let host = cpal::default_host();
+        let input = host.default_input_device()?;
+        let config = input.default_input_config().ok()?.config();
```
A bit of API work that _definitely_ needs a second opinion, but I've created a way to list and select an audio input device that is entirely separate from the `cpal` API. In addition, I've also added support for creating the audio input stream from a `cpal` device, so power-users can setup the input system using `cpal` if they want to, and just add it as a resource (wrapped in `AudioInput`)

bushrat011899 commented:
I've updated this PR to be as close to zero allocation (within the hot-path) as I think I can whilst making minimal assumptions. In addition, I've completely decoupled from `rodio` internally and for the public API, and only offer optional `cpal` APIs. For the vast majority of use cases with this API, only the Bevy controlled types will ever be interacted with, allowing for a zero-break changeover to Kira (if desired), and a minimal break if we ever moved off of `cpal` (I don't see any indication this would happen).

In addition, I have done some more latency testing and updated the description to include the details. With my limited setup, I was not able to detect a latency larger than 10ms (the floor of my measurement precision) for the entire pipeline of my example (physical sound through to visual display on my monitor). I would still encourage anyone able to try and measure this more precisely if possible. I doubt this implementation would be low enough latency for musical performance (e.g., using Bevy for custom audio effects on a MIDI instrument), but I am confident it is low enough for communication.

SolarLiner reviewed (changes requested):
I don't know why you decided against using a circular buffer (and maybe there's a valid reason, what's the stance on adding third-party crates?), but this really complexified the implementation. The whole system of reusing pre-allocated buffers is a good idea, but misuses `sync_channels` as long-term storage, and creates edge-cases when the audio input doesn't neatly fit into any one of them.

As such it's hard for me to audit the real-time safety of the code here because of its complexity, but I think in its current state, outside of the note on allocation I made, it is good.

I would really recommend using a circular buffer instead, as it would simplify a lot of the code. The `handle_input_stream` can then copy the entire unread buffer and create a single event, if going through the event system is desired.

One last point is that the implementation as currently written does not allow for using the audio input as audio data directly (ie. a `rodio::Decodable`), only for monitoring. This is fine for now, but it strikes me a feature-incomplete (though I don't know how big that use case is). It's fine to do this as another PR, though it would require sharing the audio input buffer, so it wouldn't be simple additions to this.

rudrabhoj commented:
It has been close to 4 months, any update on status of this?

SolarLiner commented:
@rudrabhoj We're in the process of reworking the audio engine entirely, as we've reached the limits of the current solution. As such this may take a while to get re-implemented, but audio input is definitely planned in the future.
